Helm Release preview announcement (#568)
pulumi/pulumi-hugo: themes/default/content/blog/announcing-kubernetes-helm-release-resource-preview/index.md

---

> title: Announcing Preview for Helm Release Resource for Kubernetes

Reviewer: ```suggestion
title: Full access to Helm features with new Helm Release resource
```

Would love to encapsulate the value this new resource enables in the title, if possible. Help grab the user's attention.

Author: Fixed.

h1: "Announcing Preview for Helm Release Resource for Kubernetes"

> date: "2021-09-07"

Reviewer: ```suggestion
date: "2021-09-15"
```

Reviewer 2: We can release the blog as soon as we are ready, no need to wait until 15th

Author: Updated to today.

> meta_desc: "Today we are announcing preview of a new Helm Release resource for Kubernetes which adds support for Helm natively on Pulumi."

Reviewer: ```suggestion
meta_desc: "Pulumi's new Helm Release resource, combines the infrastructure-as-code power of Pulumi with native support for Helm, including full access to all Helm features. Helm Release is available in public preview today."
```

Author: Updated.

> meta_image: "helm-release.gif"

Reviewer: Not sure this can be a GIF. Would be good to confirm with Christian and also confirm, if it can be a GIF, that the GIF is the correct size.

Author: > Not sure this can be a GIF. Would be good to confirm with Christian and also confirm, if it can be a GIF, that the GIF is the correct size.

The gif does render and I have seen it in some of the older Kubernetes posts. But it definitely is pretty illegible in the index view. 
@cnunciato do you have any recommendations? I am happy to switch to a generic Pulumi image like https://www.pulumi.com/blog/kubernetes-fundamentals-part-one/k8s-fundamentals.png instead.

Reviewer 3: Yeah, unfortunately the preview/meta image shouldn't be an animated GIF because social media previews won't render them consistently. See here, for example, with the [Twitter card validator](https://cards-dev.twitter.com/validator), which just shows the first frame, I believe: 

![image](https://user-images.githubusercontent.com/274700/132548175-4f9d191a-2d99-4533-bf84-5dbd0108a8e5.png)

I'd suggest going with a bigger PNG for the meta image (I put some [recommended specs](https://github.com/pulumi/pulumi-hugo/blob/master/BLOGGING.md#media) in BLOGGING.md) and maybe finding a place for the animated GIF in the body of the post.

Author: I have added a separate image - apologies in advance if it looks hideous. I used one of the sanctioned logo images from [Helm](https://github.com/cncf/artwork/blob/master/examples/graduated.md#helm-logos) on an existing k8s meta image.

authors: ["vivek-lakshmanan"]
tags: ["Kubernetes", "helm"]
---

> Kubernetes has been a significant focus of Pulumi since its very beginnings. Pulumi added support for installing [Helm charts](https://helm.sh/) way back in [2018](https://www.pulumi.com/blog/using-helm-and-pulumi-to-define-cloud-native-infrastructure-as-code/) and it has seen significant adoption by users since. However, Pulumi's current Chart integration lacks support for some increasingly common advanced features in Helm charts, e.g.:

Reviewer: The blog quote and other links within pulumi.com should likely be relrefs - see other posts for the syntax.

Author: I have updated it. Could you take a look through to make sure I miss anything? Looks like since the pulumi-hugo/docs split, the relrefs across to`/docs/` etc are not retained locally (understandably). @cnunciato - will the previews reflect them correctly? (Will check once the PR preview is rendered).

Reviewer 3: @viveklak Sorry, bit of a late response, but yes, `relref` should work both locally and in PR previews for links pointing to content that resides in this repo, so since the blog is in this repo, that'll work. (And your preview seems good.)

Author: > @viveklak Sorry, bit of a late response, but yes, `relref` should work both locally and in PR previews for links pointing to content that resides in this repo, so since the blog is in this repo, that'll work. (And your preview seems good.)

Chatted with @cnunciato offline and the relrefs to docs are expected to be broken in the preview in this repo. The merge preview on docs should render the right links.

1. Support for [Helm lifecycle hooks](https://helm.sh/docs/topics/charts_hooks/)
2. Handling sub-charts and [dependencies](https://helm.sh/docs/topics/charts/#chart-dependencies)

> These and other limitations have resulted in Helm support becoming a common painpoint for our Kubernetes users.

Reviewer 4: ```suggestion
As Helm continues to evolve, Pulumi users using the `Chart` resource have often had to get very creative in order to get the desired functionality
```

Tried to put a more positive spin on this

Author: Updated. Could you take a look to make sure you agree?

> Today we are excited to announce the *preview* of a new [Helm Release](https://www.pulumi.com/docs/reference/pkg/kubernetes/helm/v3/release/) resource starting with [v3.7.0](https://github.com/pulumi/pulumi-kubernetes/releases/tag/v3.7.0) of the Pulumi Kubernetes Provider and SDK in all Pulumi supported languages. Note this is not meant to signal deprecation of the [Chart](https://www.pulumi.com/docs/reference/pkg/kubernetes/helm/v3/chart/) resource which will continue to be supported in parallel. This new resource provides Pulumi users more options to choose the right tool for their use-case. The rest of this blog post will highlight how this resource differs from the existing Chart component resource and describe how and when to use the new resource.

Reviewer: ```suggestion
Today we are excited to announce the public preview of a new [Helm Release](https://www.pulumi.com/docs/reference/pkg/kubernetes/helm/v3/release/) resource starting with [v3.7.0](https://github.com/pulumi/pulumi-kubernetes/releases/tag/v3.7.0) of the Pulumi Kubernetes Provider and SDK in all Pulumi supported languages. This new resource provides Pulumi users more options to choose the right tool for their use-case. The rest of this blog post will highlight how this resource differs from the existing [Helm Chart](https://www.pulumi.com/docs/reference/pkg/kubernetes/helm/v3/chart/) component resource and describe how and when to use the new resource.

{{% notes %}}
The existing [Helm Chart](https://www.pulumi.com/docs/reference/pkg/kubernetes/helm/v3/chart/) resource, which extracts and converts Kubernetes resources into individual resources in the Pulumi program, will continue to be supported. 
{{% /notes %}}
```

Author: Fixed.

<!--more-->

## How is Helm Release resource different?

> Pulumi's Helm Chart integration was implemented as a [Component Resource](https://www.pulumi.com/docs/intro/concepts/resources/#components) which effectively extracts the corresponding Kubernetes resources' manifests from the chart and installs them as if they were individually specified in the Pulumi user program. This allowed users to install community developed Helm charts, while providing fine-grained control over the respective Kubernetes resources installed in the process. For instance, the installed resources are subject to all policies specified through Policy-as-Code policy packs registered with Pulumi. Other users have used Pulumi's powerful transformation support to manipulate Helm charts to fit their desired needs before installing them in their environment.

Reviewer 4: ```suggestion
Pulumi's Helm Chart integration was implemented as a [Component Resource](https://www.pulumi.com/docs/intro/concepts/resources/#components) which effectively extracts the corresponding Kubernetes resources' manifests from the chart and installs them as if they were individually specified in the Pulumi user program. This allowed users to install community developed Helm charts, while providing fine-grained control over the respective Kubernetes resources installed in the process. For instance, the installed resources are subject to all policies specified through Policy-as-Code policy packs registered with Pulumi. Other users have used Pulumi's powerful transformation support to manipulate Helm charts to fit their desired needs before installing them in their environment.
```

Author: I didn't see an actual diff in the above comment? I am also known to be blind as a bat so let me know if I missed something.

> In contrast, the new Helm Release resource is a thin wrapper around Helm itself. This provides the following advantages:

Reviewer 4: ```suggestion
In contrast, the new Helm Release resource uses the Helm SDK, bringing the full feature set of Helm directly to Pulumi users. This provides the following advantages:
```

Trying to avoid using "wrapper" as a description, it has an implication of using the Helm binary.

Author: Makes sense! Fixed.

### Helm Controls Lifecycle

> Unlike the existing Chart component resource, Helm Release directly offloads all responsibility of orchestration and lifecycle management to Helm itself. This means Helm can invoke all necessary lifecycle hooks for the release as it would through the CLI. It also updates chart dependencies and correctly handles sub-chart values etc.

Reviewer: Is there actually more that it does? If so, let's write or link to that, instead of the `etc.`

Or, if the paragraph is good as-is, let's cut the "etc."

### Configurable Await Support

> Pulumi's Helm Release support by default waits up to a configurable timeout (currently defaults to *5 minutes*) for all resources installed by the release to be available. With a sufficient timeout, the resource can be reliably passed in `dependsOn` clauses in subsequent resources to encode ordering requirements. In addition, users can also require waiting on all helm hooks to complete by setting `jobAwait` to true. However, if desired, all await behavior can be disabled by setting `skipAwait` option to true.

Reviewer: ```suggestion
By default, Pulumi's Helm Release resource waits for all resources installed by the release to be available, up to a configurable timeout that defaults to *5 minutes*. With a sufficient timeout, the resource can be reliably passed to `dependsOn` clauses in subsequent resources to encode ordering requirements. In addition, users can also require waiting on all Helm hooks to complete by setting `jobAwait` to true. However, if desired, all await behavior can be disabled by setting `skipAwait` option to true.
```

Will the audience of this blog post know why the various await options would be useful? If not, might be good to add a bit more about the value that users would get by setting `jobAwait` or `skipAwait`. Or, perhaps just a link to docs would suffice, if we have a good page to link to.

## How do I use it?

If you aren't already familiar with using Pulumi with Kubernetes, head on over to [Pulumi's getting started guide for Kubernetes](https://www.pulumi.com/docs/get-started/kubernetes/) first.

1. In your chosen Kubernetes Pulumi project, make sure the referenced version of the Pulumi-Kubernetes SDK is at least `v3.7.0`.
2. Helm Release support is configured to target the relevant cluster pointed to by the Kubernetes provider associated with it and no additional configuration is required by default. However, more advanced configuration options are supported, see the provider configuration [documentation](https://www.pulumi.com/docs/reference/pkg/kubernetes/provider/#helmdriver_nodejs) for more details.
3. Refer to the instructions for your chosen language in the [resource documentation](https://www.pulumi.com/docs/reference/pkg/kubernetes/helm/v3/release/#create) to create a Helm Release resource.

Lets look at a concrete example of Helm Release in action. In the following snippet we install [Redis](https://redis.io/) through a Helm Chart in each of the supported Pulumi languages using the new Helm Release resource:

> {{< chooser language "typescript,go,python" >}}

Reviewer 2: A reminder to add  C#

Author: Done now.

const pass = new random.RandomPassword("pass", {length:10});
const redisPassword = pass.result;

const namespace = new k8s.core.v1.Namespace("redis-ns");

const release = new k8s.helm.v3.Release("redis-helm", {
    chart: "redis",
    repositoryOpts: {
        repo: "https://charts.bitnami.com/bitnami",
    },
    version: "13.0.0",
    namespace: namespace.metadata.name,
    // Values from Chart's parameters specified hierarchically,
    // see https://artifacthub.io/packages/helm/bitnami/redis/13.0.0#parameters for reference.
    values: {
        cluster: {
            enabled: true,
            slaveCount: 3,
        },
        metrics: {
            enabled: true,
            service: {
                annotations: {
                    "prometheus.io/port": "9127",
                }
            },
        },
        global: {
            redis: {
                password: redisPassword,
            }
        },
        rbac: {
            create: true,
        }
    },
    // By default Release resource will wait till all created resources
    // are available. Set this to true to skip waiting on resources being
    // available.
    skipAwait: false,
});

> // We can lookup resources once the release is installed. The releases'

Reviewer: ```suggestion
// We can lookup resources once the release is installed. The release's
```

    pulumi.interpolate`${release.status.namespace}/${release.status.name}-redis-master`);
export const redisMasterClusterIP = srv.spec.clusterIP;
```

{{% /choosable %}}

{{% choosable language go %}}

```go
package main

import (
	"fmt"

	corev1 "github.com/pulumi/pulumi-kubernetes/sdk/v3/go/kubernetes/core/v1"
	"github.com/pulumi/pulumi-kubernetes/sdk/v3/go/kubernetes/helm/v3"
	"github.com/pulumi/pulumi-random/sdk/v4/go/random"
	"github.com/pulumi/pulumi/sdk/v3/go/pulumi"
)

func main() {
	pulumi.Run(func(ctx *pulumi.Context) error {

		namespace, err := corev1.NewNamespace(ctx, "redis-ns", nil)
		if err != nil {
			return err
		}

		redisPassword, err := random.NewRandomPassword(ctx, "pass", &random.RandomPasswordArgs{
			Length: pulumi.Int(10),
		})
		if err != nil {
			return err
		}

		rel, err := helm.NewRelease(ctx, "redis-helm", &helm.ReleaseArgs{
			Chart: pulumi.String("redis"),
			RepositoryOpts: helm.RepositoryOptsArgs{
				Repo: pulumi.String("https://charts.bitnami.com/bitnami"),
			},
			Version:   pulumi.String("13.0.0"),
			Namespace: namespace.Metadata.Name(),

      // Values from Chart's parameters specified hierarchically,

>       // see https://artifacthub.io/packages/helm/bitnami/redis/13.0.0#parameters for reference.

Reviewer 2: Indentation looks off here

Author: Fixed now.

			Values: pulumi.Map{
				"cluster": pulumi.Map{
					"enabled": pulumi.Bool(true),
				},
				"metrics": pulumi.Map{
					"enabled": pulumi.Bool(true),
					"service": pulumi.Map{
						"annotations": pulumi.StringMap{
							"prometheus.io/port": pulumi.String("9127"),
						},
						"type": pulumi.String("ClusterIP"),
					},
				},
				"global": pulumi.Map{
					"redis": pulumi.Map{
						"password": redisPassword,
					},
				},
				"rbac": pulumi.BoolMap{
					"create": pulumi.Bool(true),
				},
			},

			// By default Release resource will wait till all created resources
			// are available. Set this to true to skip waiting on resources being
			// available.
			SkipAwait: pulumi.BoolPtr(false),
		})
		if err != nil {
			return err
		}

> 		// We can lookup resources once the release is installed. The releases'

Reviewer: ```suggestion
		// We can lookup resources once the release is installed. The release's
```

		// status field is set once the installation completes, so this, combined
		// with `skipAwait: pulumi.BoolPtr(false)` above, will wait to retrieve

>     //  the Redis master ClusterIP till all resources in the Chart are available.

Reviewer 2: and here

Author: Fixed.

		svc := pulumi.All(rel.Status.Namespace(), rel.Status.Name()).
			ApplyT(func(r interface{}) (interface{}, error) {
				arr := r.([]interface{})
				namespace := arr[0].(*string)
				name := arr[1].(*string)
				svc, err := corev1.GetService(ctx,
					"redis-master-svc",
					pulumi.ID(fmt.Sprintf("%s/%s-redis-master", *namespace, *name)),
					nil,
				)
				if err != nil {
					return "", nil
				}
				return svc.Spec.ClusterIP(), nil
			})
		ctx.Export("redisMasterClusterIP", svc)

		return nil
	})
}

```

{{% /choosable %}}

{{% choosable language python %}}

```python
import pulumi
from pulumi import Output
from pulumi_random.random_password import RandomPassword
from pulumi_kubernetes.core.v1 import Namespace, Service
from pulumi_kubernetes.helm.v3 import Release, ReleaseArgs, RepositoryOptsArgs

namespace = Namespace("redis-ns")

redis_password = RandomPassword("pass", length=10)
release = Release("redis-helm",

>                   args=ReleaseArgs(

Reviewer 2: Can we format this code so that there's less space at the left and more room for the useful code?

Author: Updated.

                      # for reference.
                      values={
                          "cluster": {
                              "enabled": True,
                              "slaveCount": 3,
                          },
                          "metrics": {
                              "enabled": True,
                              "service": {
                                  "annotations": {
                                      "prometheus.io/port": "9127",
                                  }
                              },
                          },
                          "global": {
                              "redis": {
                                  "password": redis_password,
                              }
                          },
                          "rbac": {
                              "create": True,
                          },
                      },
                  ),
                  # By default Release resource will wait till all created resources
                  # are available. Set this to true to skip waiting on resources being
                  # available.
                  skip_await=False)

> # We can lookup resources once the release is installed. The releases'

Reviewer: ```suggestion
# We can lookup resources once the release is installed. The release's
```

# status field is set once the installation completes, so this, combined
# with `skip_await=False` above, will wait to retrieve the Redis master
# ClusterIP till all resources in the Chart are available.
status = release.status
srv = Service.get("redis-master-svc",
                  Output.concat(status.namespace, "/", status.name, "-redis-master"))
pulumi.export("redisMasterClusterIP", srv.spec.cluster_ip)

```

{{% /choosable %}}

{{< /chooser >}}

Users of the existing Helm Chart resource will notice that we retained the hierarchical `values` specification experience and a similar means to configure the repository. The above example also highlights the await and resource lookup experience.

> ![helm-release](./helm-release.gif)

Reviewer: ```suggestion
![helm-release](helm-release.gif)
```

I think the `./` are unneeded. Also, it would be good to write meta text within the square brackets describing what's shown in the GIF; that helps folks using screen readers to understand what they can't see.

## Which Helm resource to use and when?

> In summary, if your primary goal is to install Helm charts with configuration options natively supported by the chart, then Helm  Release should be a good fit. If on the other hand you intend to manipulate Helm Charts through transformations (due to bugs in the charts or restrictions in your environment), then the experience is much richer with the Helm Chart resource.

Reviewer: ```suggestion
In summary, if your primary goal is to install Helm charts with configuration options natively supported by the chart, then Helm Release should be a good fit. If on the other hand you intend to manipulate Helm Charts through transformations (due to bugs in the charts or restrictions in your environment), then the experience is much richer with the Helm Chart resource.
```

Author: Fixed.

If you intend to use Policy-as-Code to enforce resource policies at a Kubernetes resource granularity, then the Helm Chart component resource would be the preferred model to use.

Note that both resources can co-exist safely in the same program.

## What's next?

> The new Helm Release resource is available in *preview* now. During the preview the resource may evolve rapidly and users are requested to not use it for critical infrastructure. We have already received valuable feedback to shape this resource and we look forward to hearing more from our users. Please reach out by filing or commenting on issues on [Github](https://github.com/pulumi/pulumi-kubernetes/issues) and reaching out to us on [Community Slack](https://slack.pulumi.com). Happy charting!

Reviewer 2: Nit: "reach out" is used twice in the last sentence

Author: Fixed.
